File: web/frontend/src/lib/api/workflow.ts
/** 工作流域 API — 运行列表/详情/启动/停止/续跑。 */

import { api } from "./client";

export interface WorkflowRun {
  run_id: string;
  name: string;
  status: "running" | "completed" | "failed" | "stopped";
  stop_reason?: string;
  scope?: string;
  parent_run_id?: string;
  spec_hash?: string;
  created_at?: number;
  finished_at?: number | null;
  running?: boolean;
}

export interface WorkflowNode {
  key: string;
  ordinal: number;
  kind: "ask" | "tool" | "repair";
  status: "running" | "completed" | "failed";
  result_preview: string;
  error: string;
  delegation_id: string;
  created_at: number;
  updated_at: number;
}

export interface WorkflowEvent {
  sequence: number;
  type: string;
  payload: Record<string, unknown>;
  ts: number;
}

export interface WorkflowRunDetail {
  run: WorkflowRun;
  nodes: WorkflowNode[];
  events: WorkflowEvent[];
}

export const workflowApi = {
  listRuns: (limit = 30) =>
    api.get<{ runs: WorkflowRun[] }>("/workflow/runs", { params: { limit } }),
  runDetail: (runId: string) =>
    api.get<WorkflowRunDetail>(`/workflow/runs/${encodeURIComponent(runId)}`),
  startRun: (spec: unknown, resumeOf = "") =>
    api.post<WorkflowRun>("/workflow/runs", { spec, resume_of: resumeOf }),
  stopRun: (runId: string) =>
    api.post<{ ok: boolean; note?: string }>(
      `/workflow/runs/${encodeURIComponent(runId)}/stop`,
    ),
  resumeRun: (runId: string) =>
    api.post<WorkflowRun>(`/workflow/runs/${encodeURIComponent(runId)}/resume`),
};

File: web/routers/workflow.py
"""工作流 API 路由。"""

from __future__ import annotations

from typing import Any, Dict

from fastapi import APIRouter, HTTPException
from pydantic import BaseModel

from services.workflow import (
    WorkflowRuntimeError,
    WorkflowService,
    WorkflowServiceError,
)

router = APIRouter(prefix="/workflow", tags=["workflow"])

_wf_svc = WorkflowService()


class StartRunRequest(BaseModel):
    """启动请求：spec 为工作流规格对象；resume_of 为修订源 run_id。"""

    spec: Dict[str, Any]
    scope: str = ""
    resume_of: str = ""


@router.get("/runs")
async def list_runs(limit: int = 30) -> Dict[str, Any]:
    return {"runs": await _wf_svc.list_runs(limit)}


@router.post("/runs")
async def start_run(req: StartRunRequest) -> Dict[str, Any]:
    try:
        return await _wf_svc.start(req.spec, scope=req.scope, resume_of=req.resume_of)
    except WorkflowRuntimeError as e:
        raise HTTPException(503, str(e)) from e
    except WorkflowServiceError as e:
        raise HTTPException(400, str(e)) from e


@router.get("/runs/{run_id}")
async def run_detail(run_id: str) -> Dict[str, Any]:
    try:
        return await _wf_svc.run_detail(run_id)
    except WorkflowRuntimeError as e:
        raise HTTPException(503, str(e)) from e
    except WorkflowServiceError as e:
        raise HTTPException(404, str(e)) from e


@router.post("/runs/{run_id}/stop")
async def stop_run(run_id: str) -> Dict[str, Any]:
    try:
        return _wf_svc.stop(run_id)
    except WorkflowRuntimeError as e:
        raise HTTPException(503, str(e)) from e
    except WorkflowServiceError as e:
        raise HTTPException(400, str(e)) from e


@router.post("/runs/{run_id}/resume")
async def resume_run(run_id: str) -> Dict[str, Any]:
    try:
        return await _wf_svc.resume(run_id)
    except WorkflowRuntimeError as e:
        raise HTTPException(503, str(e)) from e
    except WorkflowServiceError as e:
        raise HTTPException(400, str(e)) from e
